ic-utils/src/call.rs

```diff
@@ -132,37 +147,42 @@ where
 }
 
 /// A synchronous call encapsulation.
-#[derive(Clone)]
-pub struct SyncCaller<'agent> {
-    agent: &'agent Agent,
-    canister_id: Principal,
-    method_name: String,
-    arg: Vec<u8>,
-    expiry: Expiry,
+pub struct SyncCaller<'agent, Out>
+where
+    Out: for<'de> ArgumentDecoder<'de> + Send + Sync,
+{
+    pub(crate) agent: &'agent Agent,
+    pub(crate) canister_id: Principal,
+    pub(crate) method_name: String,
+    pub(crate) arg: Result<Vec<u8>, AgentError>,
+    pub(crate) expiry: Expiry,
+    pub(crate) phantom_out: std::marker::PhantomData<Out>,
 }
 
-impl<'agent> SyncCaller<'agent> {
+impl<'agent, Out> SyncCaller<'agent, Out>
+where
+    Out: for<'de> ArgumentDecoder<'de> + Send + Sync,
+{
     /// Perform the call, consuming the the abstraction.
-    async fn call<R>(self) -> Result<R, AgentError>
-    where
-        R: for<'de> ArgumentDecoder<'de> + Send + Sync,
-    {
-        self.agent
-            .query(&self.canister_id, &self.method_name)
-            .with_arg(&self.method_name)
-            .call()
-            .await
-            .and_then(|r| decode_args(&r).map_err(|e| AgentError::CandidError(Box::new(e))))
+    async unsafe fn call_raw(self) -> Result<Vec<u8>, AgentError> {
```

Comment:
Is the whole reason for `unsafe` here to discourage direct calls to call_raw?

It seems like as a side-effect, this function could accumulate unsafe code.

Could we change this function to a one-liner that forwards to a function that is not marked `unsafe` ?

Reply:
Or is there some other way to hide `call_raw` ?

Reply:
Feature flag would be another way. You have to "opt in" at the crate level, which is fine I guess.

Reply:
Done. PTAL